frame/contracts/src/migration/v15.rs

```diff
@@ -0,0 +1,330 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Move contracts' _reserved_ balance from the `deposit_account` to be _held_ in the contract's
+//! account instead. Since [`Currency`](frame_support::traits::Currency) has been
+//! [deprecated](https://github.com/paritytech/substrate/pull/12951), we need the deposits to be
+//! handled by the [`frame_support::traits::fungible`] traits instead. For this transfer the
+//! balance in the deposit account to the contract's account and hold it in there.
+//! Then the deposit account is not needed anymore and we can get rid of it.
+
+use crate::{
+	migration::{IsFinished, MigrationStep},
+	weights::WeightInfo,
+	AccountIdOf, BalanceOf, CodeHash, Config, HoldReason, Pallet, TrieId, Weight, LOG_TARGET,
+};
+#[cfg(feature = "try-runtime")]
+use frame_support::{dispatch::Vec, traits::fungible::InspectHold};
+use frame_support::{
+	pallet_prelude::*,
+	storage_alias,
+	traits::{
+		fungible::{Mutate, MutateHold},
+		tokens::{fungible::Inspect, Fortitude, Preservation},
+	},
+	BoundedBTreeMap, DefaultNoBound,
+};
+use frame_system::Pallet as System;
+use sp_core::hexdisplay::HexDisplay;
+#[cfg(feature = "runtime-benchmarks")]
+use sp_runtime::traits::{Hash, TrailingZeroInput};
+#[cfg(feature = "try-runtime")]
```

Comment:
Let's move it into the function or use fully qualified names to reduce the amount of conditional compilation attributes.